controlplane/src/core/repositories/FederatedGraphRepository.ts

```diff
@@ -782,4 +788,188 @@ export class FederatedGraphRepository {
+
+    // To avoid depending on postgres, we do a second query to get the inserted client
+    const result = await this.db.query.federatedGraphClients.findFirst({
+      columns: { id: true },
+      where: and(
+        eq(federatedGraphClients.name, clientName),
+        eq(federatedGraphClients.federatedGraphId, graph.federatedGraph.id),
+      ),
+    });
+    return result!.id;
+  }
+
+  public async getClients(federatedGraphName: string): Promise<ClientDTO[]> {
+    const userRepo = new UserRepository(this.db);
+
+    const graph = await this.db.query.targets.findFirst({
+      columns: {},
+      with: {
+        federatedGraph: {
+          columns: { id: true },
+        },
+      },
+      where: and(
+        eq(schema.targets.name, federatedGraphName),
+        eq(schema.targets.organizationId, this.organizationId),
+        eq(schema.targets.type, 'federated'),
+      ),
+    });
+    if (graph === undefined) {
+      throw new Error(`could not find graph ${federatedGraphName}`);
+    }
+
+    const fedGraphClients = await this.db
+      .select({
+        name: federatedGraphClients.name,
+        id: federatedGraphClients.id,
+        createdAt: federatedGraphClients.createdAt,
+        createdBy: federatedGraphClients.createdBy,
+        lastUpdatedAt: federatedGraphClients.updatedAt,
+        lastUpdatedBy: federatedGraphClients.updatedBy,
+      })
+      .from(federatedGraphClients)
+      .where(eq(federatedGraphClients.federatedGraphId, graph.federatedGraph.id));
+
+    const clients: ClientDTO[] = [];
+
+    for (const c of fedGraphClients) {
+      const createdUser = await userRepo.byId(c.createdBy);
+      let updatedUser: UserDTO | null = null;
+      if (c.lastUpdatedBy) {
+        updatedUser = await userRepo.byId(c.lastUpdatedBy);
+      }
+
+      clients.push({
+        id: c.id,
+        name: c.name,
+        createdAt: c.createdAt.toISOString(),
+        lastUpdatedAt: c.lastUpdatedAt?.toISOString() || '',
+        createdBy: createdUser?.email || '',
+        lastUpdatedBy: updatedUser ? updatedUser.email : '',
+      });
+    }
+
+    return clients;
+  }
+
+  public async updatePersistedOperations(federatedGraphName: string, userId: string, clientId: string, operations: {hash: string, filePath: string;}[]) {
+    const graph = await this.db.query.targets.findFirst({
+      columns: {},
+      with: {
+        federatedGraph: {
+          columns: { id: true },
+        },
+      },
+      where: and(
+        eq(schema.targets.name, federatedGraphName),
+        eq(schema.targets.organizationId, this.organizationId),
+        eq(schema.targets.type, 'federated'),
+      ),
+    });
+    if (graph === undefined) {
+      throw new Error(`could not find graph ${federatedGraphName}`);
+    }
+    const now = new Date()
+    for (const operation of operations) {
+      await this.db.insert(federatedGraphPersistedOperations).values({
```

Comment:
You can batch insert. Map the operations to an array.

Reply:
Can we really? Notice there's a `onConflictDoUpdate()` - I've found this batching API but it seems it's only supported with 2 specific drivers https://orm.drizzle.team/docs/batch-api